core/raftlease/fsm.go

```diff
@@ -518,6 +536,79 @@ func (f *FSM) Restore(reader io.ReadCloser) error {
 	return nil
 }
 
+// BatchFSM creates a FSM that allows for batching operations. Raft takes
+// care of applying the batches in chunked sizes, which allows for restoring
+// and snapshotting a the library level. Those should be transparent to the
+// the FSM.
+type BatchFSM struct {
+	*FSM
+}
+
+// NewBatchFSM creates a BatchFSM from an existing FSM. By lifting the FSM
+// into a BatchFSM allows for better performance when applying logs, it does
+// this by only stealing a lock only when required.
+func NewBatchFSM(fsm *FSM) *BatchFSM {
+	return &BatchFSM{
+		FSM: fsm,
+	}
+}
+
+type batchValue struct {
+	Command  Command
+	Response *response
+}
+
+// ApplyBatch is invoked once a batch of log entries has been committed and
+// are ready to be applied to the FSM. ApplyBatch will take in an array of
+// log entries. These log entries will be in the order they were committed,
+// will not have gaps, and could be of a few log types. Clients should check
+// the log type prior to attempting to decode the data attached. Presently
+// the LogCommand and LogConfiguration types will be sent.
+//
+// The returned slice must be the same length as the input and each response
+// should correlate to the log at the same index of the input. The returned
+// values will be made available in the ApplyFuture returned by Raft.Apply
+// method if that method was called on the same Raft node as the FSM.
+// ApplyBatch is part of raft.BatchingFSM.
+func (f *BatchFSM) ApplyBatch(logs []*raft.Log) interface{} {
+	// Unmarshal all the logs up front, we can validate them without
+	// stealing the lock. Additionally we can ensure that we get the correct
+	// type of log.
+	values := make([]batchValue, len(logs))
+	for i, log := range logs {
+		command, err := unmarshalCommand(log)
+		if err != nil {
+			values[i] = batchValue{
+				Response: &response{err: errors.Trace(err)},
+			}
+			continue
+		}
+
+		values[i] = batchValue{
+			Command: command,
+		}
+	}
+
+	// Ensure we steal the lock so we serialize all log applications. That way
+	// we never allow another Apply or ApplyBatch to intervene.
+	f.mu.Lock()
+	defer f.mu.Unlock()
+
+	responses := make([]interface{}, len(values))
```

Comment:
This allocation could be performed without holding the lock. In fact, it can be pre-allocated at the top of the method and the unmarhaling code could populate errors directly (also see comment above).